[영화 극장 선택] 채넛 1, 2단계 미션 제출합니다.

yrsel opened the pull request:
# 셀프 체크리스트

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [ ] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [ ] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

 <!--
 학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
 정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
 -->
 
안드로이드 코드에서 저의 악습관, 도메인 설계의 문제에 대해 많이 리뷰 부탁드립니다 :)

## 코드 리뷰 커뮤니케이션

안녕하세요, 페로로 🙇‍♂️
영화 극장 선택 미션 잘 부탁드립니다 ! 
이전 미션을 진행하며 안드로이드에 대한 이해도가 많이 부족함을 느끼고 시간 관리의 중요성을 느꼈습니다.
이번 미션에서는 조금 더 선택과 집중을 잘 해보려고 합니다 ! 

도메인 설계에 있어서 영화극장들, 영화극장, 영화 스케줄, 영화 잘게 쪼개었고, 상태를 외부에 최대한 노출 시키지 않게 않도록 일급컬렉션을 활용하였습니다. 그러다보니 연쇄 호출이 발생하게 되고 이해하기가 어려워진 느낌이 있습니다.
이러한 부분에서 '도메인 설계에 문제가 있구나' 하는 생각이 들었습니다. 도메인 설계에 대해 코멘트 부탁드립니다. ! 

아직 안드로이드에 익숙하지 않아서 안드로이드에서 어떤 부분에서 실수하는 지 조차 인식을 못하는 경우가 많이 있는 것 같습니다.
고쳤으면 하는 부분에 대해 많이 많이 언급해주셨으면 좋겠습니다.
이번 미션도 잘 부탁드립니다, 감사합니다 :-)

## 스크린샷

- 시연 영상

https://github.com/user-attachments/assets/e0ad4b43-7b81-4ab7-9277-b425ad675730

ijh1298 committed:
build: 기본 코드 준비

ijh1298 committed:
chore: 불필요한 파일 제거

ijh1298 committed:
docs: README 기능 요구 사항 작성

yrsel committed:
feat: 상영 정보를 나타내는 MovieSchedule 추가

yrsel committed:
feat: 극장을 나타내는 Theater 추가

yrsel committed:
chore: DataBinding gradle 설정 추가

yrsel committed:
feat: MovieItem에 DataBinding 적용

yrsel committed:
chore: ktlint format 적용

ijh1298 committed:
chore: Project 이름 수정

ijh1298 committed:
feat: BookingActivity에 DataBinding 적용

yrsel committed:
refactor: 더미 데이터 분리

yrsel committed:
refactor: ScreeningDate -> ScreeningPeriod 변경

yrsel committed:
feat: 영화가 상영중인 극장들을 가져오는 기능 구현

ijh1298 committed:
chore: Fragment 의존성 추가

ijh1298 committed:
feat: TheaterBottomSheetDialogFragment 추가

ijh1298 committed:
feat: theater_item 추가

ijh1298 committed:
feat: MovieListActivity 내 DataBinding 적용 및 Fragment 추가

yrsel committed:
chore: 의존성 Notation 수정

yrsel committed:
feat: BookingActivity intent 데이터 Movie -> Theater 변경

yrsel committed:
feat: TheaterBottomSheetDialogFragment 에 RecyclerView 구현

yrsel committed:
feat: 도메인 직렬화 추가

yrsel committed:
chore: 더미 데이터 수정

yrsel committed:
refactor: 상수 분리

ijh1298 committed:
feat: TheaterBottomSheetDialogFragment에 Movie 데이터 전달

ijh1298 committed:
refactor: 상영 가능한 극장 메서드 기본 인자 추가

ijh1298 committed:
chore: 더미 데이터 수정

ijh1298 committed:
refactor: 날짜, 시간 스피너 업데이트 로직 수정

ijh1298 committed:
refactor: 상영 중인지 확인하는 로직 수정

ijh1298 committed:
chore: KtlintFormat 적용

ijh1298 committed:
refactor: 상영 가능 영화가 없을 경우 상영관 클릭 제한 추가

yrsel committed:
refactor: 극장 이름이 길 경우 marquee 효과 추가

yrsel committed:
refactor: BookedTicket 극장 이름 프로퍼티 추가

yrsel committed:
refactor: BookingSeatPresenter 에 Theater 상태 정보 추가

yrsel committed:
test: ScreeningDateTime 테스트 추가

yrsel committed:
test: 테스트 코드 수정

yrsel committed:
refactor: movieScheduleByMovie -> theaterByMovie 함수명 수정

yrsel committed:
chore: KtlintFormat 적용

yrsel committed:
test: TheaterTest 테스트 추가

- Fixture 추가

yrsel committed:
test: 테스트 더미 데이터 수정

yrsel committed:
test: TheatersTest 테스트 추가

ijh1298 committed:
feat: MovieBookingActivity 추가

ijh1298 committed:
feat: MovieBookingActivity 레이아웃 추가

yrsel committed:
refactor: MovieListActivity -> MovieListFragment 수정

yrsel committed:
feat: MovieBookingActivity에 fragment 연결

yrsel committed:
feat: MovieBookingActivity에 예매내역, 설정 fragment 연결

yrsel committed:
chore: kapt 플러그인 추가

yrsel committed:
feat: BindingAdapters 이미지 불러오기 메서드 추가

yrsel committed:
chore: KtlintFormat 적용

yrsel committed:
refactor: BookingActivity 영화 포스터 BindingAdapter 적용

yrsel committed:
refactor: BookingCompleteActivity에서 뒤로가기 대상 Activity 변경

yrsel committed:
refactor: TheaterBottomSheetDialogFragment 클릭 시, remove 호출 추가

yrsel committed:
test: MovieBookingActivityTest UI 테스트 추가

yrsel committed:
test: MovieListFragmentTest 테스트 코드 수정

yrsel committed:
test: BookingSeatActivityTest 테스트 코드 수정

yrsel committed:
test: BookingCompleteActivityTest 테스트 코드 수정

yrsel committed:
refactor: TheaterBottomSheetDialogFragment MVP 패턴 적용

yrsel committed:
test: TheaterBottomSheetDialogPresenterTest 테스트 추가

laco-dev commented on app/src/main/java/woowacourse/movie/domain/model/ScreeningDateTime.kt:
```diff
@@ -0,0 +1,20 @@
+import java.time.LocalDate
+import java.time.LocalTime
+
+class ScreeningDateTime(
+    val screeningDate: LocalDate,
+    val screeningTime: LocalTime,
+) : Serializable {
+    fun isScreening(
+        selectedDate: LocalDate,
+        selectedTime: LocalTime,
+    ): Boolean =
+        screeningDate.isAfter(selectedDate) ||
+            (
+                screeningDate.isEqual(selectedDate) &&
+                    screeningTime.isAfter(selectedTime)
+            )
+}
```
이 객체의 역할이 이해가 잘 안됩니다.
논리적으로 당일 밤 11시 59분을 넣어도 상영 가능하다는 값이 반환되지 않나요?

어떤 영화의 상영 시간을 담는 값 객체같은데 실제로 상영 가능한지를 비교하려면 더 많은 정보가 필요한거 아닌가요?
주말, 평일, 마지막 상영시간 등등

단지 현재 시간 이후라면 무조건 상영 가능하다. 를 표현하고 싶었던 것일까요?

yrsel replied:
> 단지 현재 시간 이후라면 무조건 상영 가능하다. 를 표현하고 싶었던 것일까요?

넵, 원래 의도는 '현재 시간 이후라면 무조건 상영 가능하다'를 표현하려고 했던 것이 맞습니다 

그러나 도메인 설계가 어색하게 느껴져 해당 클래스 제거하게 되었습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/domain/model/ScreeningTimeItems.kt:
```diff
@@ -0,0 +1,25 @@
+import java.time.LocalDateTime
+import java.time.LocalTime
+
+class ScreeningTimeItems(
+    val screeningTimes: List<LocalTime> = emptyList(),
+) {
```
도메인 모델에 Items와 같은 이름은 좋은 표현이 아닙니다.
예를 들어 다음 코드는 어떻게 해석해야 할까요?
```kotlin
val timeItemsList: List<ScreeingTimeItems> = ...
```

아이템이 뭘 기준으로 아이템이라는 것인지도 구분이 어렵기도 하고 클래스로 굳이 분리될 필요성이 있는지도 조금 고민이 듭니다.

yrsel replied:
제가 봐도 무슨 역할을 하는지 단번에 이해하기 어렵고 역할이 모호하다고 생각되었습니다.
설계를 다시 하며 해당 클래스 제거하게 되었습니다

laco-dev commented on app/src/main/java/woowacourse/movie/domain/model/MovieSchedule.kt:
```diff
@@ -0,0 +1,19 @@
+import java.time.LocalDate
+import java.time.LocalTime
+
+class MovieSchedule(
+    val movie: Movie,
+    val screeningDateTime: ScreeningDateTime,
+    val seat: Seats,
+) : Serializable {
```
영화 일정에 좌석은 왜 있나요?
좌석으로부터 무엇을 알고 싶은것인지 궁금합니다.

yrsel replied:
영화 일정마다 예매된 기존 좌석이 있을 수 있다고 가정하여 기존 좌석은 선택하지 못하게 하면서 좌석을 선택할 수 있는 것을 고려했습니다.
하지만, 현재 1 ~ 4 단계 요구사항에서 해당 기능은 필요없는 것 같습니다.😅

laco-dev commented on app/src/main/java/woowacourse/movie/domain/model/MovieSchedule.kt:
```diff
@@ -0,0 +1,19 @@
+    val screeningDateTime: ScreeningDateTime,
+    val seat: Seats,
+) : Serializable {
+    fun isScreeningByDateAndTime(
+        selectedMovie: Movie,
+        selectedDate: LocalDate,
+        selectedTime: LocalTime,
+    ): Boolean {
+        return movie == selectedMovie && screeningDateTime.isScreening(selectedDate, selectedTime)
+    }
```
이미 결정된 영화 정보에게 상영이 가능한지 묻는건데 선택한 영화 정보는 왜 넘기는거에요?
영화들은 영화들대로 있고, 영화 일정도 일정대로 따로따로 노는건가요?

제 생각으로는 아래 함수 하나로 동일한 역할을 하는것은 아닌가 싶어서요.
```kotlin
fun isScreening(dateTime: LocalDateTime): Boolean { ... }
```

yrsel replied:
도메인 설계를 변경하면서 영화 데이터를 받지 않도록 수정하였습니다 !

laco-dev commented on app/src/main/java/woowacourse/movie/domain/model/Theater.kt:
```diff
@@ -0,0 +1,23 @@
+    val name: String,
+    val movieSchedules: List<MovieSchedule>,
+) : Serializable {
+    fun theaterByMovie(
+        movie: Movie,
+        date: LocalDate,
+        time: LocalTime,
+    ): Theater {
+        val filtered: List<MovieSchedule> =
+            movieSchedules.filter { movieSchedule ->
+                movieSchedule.isScreeningByDateAndTime(movie, date, time)
+            }
+
+        return Theater(name, filtered)
+    }
```
특정 영화에 대한 상영관을 찾으려면

최초 모든 영화 정보를 담고 있는 상영관으로부터 계속해서 상영관을 필터링해서 만들어내야 하나요?
그러면 논리적으로 A 상영관이라는 이름이 서로 다른 영화들을 방영해도 어색하지 않다고 보시나요?

필터링을 할 때 마다 원본 상영관에서 방영하는 영화들이 계속해서 오염되고 있어요.

yrsel replied:
영화관은 원본 데이터를 유지하면서 영화에 대한 정보가 들어온다면 영화관이 해당 영화에 대한 스케줄을 반환할 수 있는 구조로 변경하였습니다 !

laco-dev commented on app/src/main/java/woowacourse/movie/domain/model/Theaters.kt:
```diff
@@ -0,0 +1,21 @@
+class Theaters(
+    val theaters: List<Theater>,
+) : Serializable {
+    fun availableTheaters(
+        movie: Movie,
+        date: LocalDate = LocalDate.now(),
+        time: LocalTime = LocalTime.now(),
+    ): Theaters {
```
이거 계속해서 반복되는데 LocalDateTime으로 하는 대신 매번 date time 따로 움직이는 이유가 뭔가요?

yrsel replied:
이전 설계에서는 스피너 데이터를 생각하고 `LocalDate`와 `LocalTime`으로 분리했었습니다.
다시 생각해보니 굳이 나눠서 전달할 필요가 없을 것 같아 페로로의 의견대로 `LocalDateTime`으로 전달하는 방식으로 변경하였습니다 !

laco-dev commented on app/src/main/java/woowacourse/movie/ui/MovieBookingActivity.kt:
```diff
@@ -0,0 +1,82 @@
+        applyWindowInsets()
+
+        // 앱 초기 실행 시 홈화면으로 설정
+        if (savedInstanceState == null) {
+            supportFragmentManager.commit {
+                setReorderingAllowed(true)
+                replace(R.id.main_fragment_container_view, MovieListFragment())
+            }
+        }
+        binding.navigation.selectedItemId = R.id.navigation_home
```
savedInstanceState 에 대한 조건문을 왜 사용하게 되었는지 설명해주실 수 있나요?
그리고 나서 selectedItemId를 홈으로 고정하면 어떤 현상이 생기는지도 알려주세요.

yrsel replied:
`savedInstanceState` 에 대한 null 조건문을 사용하지 않을 경우 configuration change 발생 할 경우 사용자가 보고 있는 화면이 무엇이었든 간에 `MovieListFragment`로 replace 되는 문제가 발생합니다.

selectedItemId를 홈으로 고정하게 되는 것 또한 configuration change, 프래그먼트가 다시 그려지는 과정에서 바텀 네비게이션은 항상 홈을 가리키게 되는 문제가 발생하기에 null이 아닐 경우에는 최상위 프래그먼트와 연관된 바텀 네비게이션이 선택될 수 있도록 해야합니다.

관련 커밋 c270950d6f395a4dd72790ae9bd4e9c1533a3cd0

laco-dev commented on app/src/main/java/woowacourse/movie/ui/MovieBookingActivity.kt:
```diff
@@ -0,0 +1,82 @@
+
+    private fun setBottomNavigationView() {
+        binding.navigation.setOnItemSelectedListener { item ->
+            when (item.itemId) {
+                R.id.navigation_home -> {
+                    supportFragmentManager.commit {
+                        setReorderingAllowed(true)
+                        replace(R.id.main_fragment_container_view, MovieListFragment())
+                    }
+                    true
+                }
+
+                R.id.navigation_history -> {
+                    supportFragmentManager.commit {
+                        setReorderingAllowed(true)
+                        replace(R.id.main_fragment_container_view, BookingHistoryFragment())
+                    }
+                    true
+                }
+
+                R.id.navigation_settings -> {
+                    supportFragmentManager.commit {
+                        setReorderingAllowed(true)
+                        replace(R.id.main_fragment_container_view, SettingsFragment())
+                    }
+                    true
+                }
+
+                else -> false
+            }
+        }
```
화면을 이동할 때 마다 프래그먼트 인스턴스가 매번 새로 만들어지지 않을까요?

yrsel replied:
프래그먼트 인스턴스들을 지연초기화하여 사용될 때 1번만 생성될 수 있도록 수정하였습니다.

관련 커밋 b28e3b3a7012aa35773fa289a1c904114c675b00

laco-dev commented on app/src/main/java/woowacourse/movie/ui/booking/contract/BookingContract.kt:
```diff
@@ -32,7 +33,7 @@ interface BookingContract {
 
     interface View {
-        fun getMovie(): Movie?
+        fun getTheater(): Theater?
```
지난 미션에서 피드백을 못받으셨을거라 생각하고 남겨드립니다.

MVP 패턴에서는 View와 Presenter간의 단방향으로 요청을 하기 때문에 무언가를 "반환" 하지 않습니다.

```
View -> Presenter
Presenter -> View
```
의 흐름대로 필요한 요청을 전달하도록 만들어주세요

yrsel replied:
MVP 패턴의 목적에 맞게 `View`와 `Presenter`의 단방향 소통을 할 수 있도록 하였고 반환값 모두 제거하였습니다

관련 커밋 
- MovieListFragment (ee0f4aeddc4670ab9fe117cfd2e3b336a6756885)
- TheaterBottomSheetDialogFragment] (14341ca6799a70c3cd8039688ed0fe6f95d1d804)
- BookingSeatActivity (ad28b44efc454126ac370d5e2692ad5930c95881)
- BookingCompleteActivity (25676b20a6730570bfbda8fed43e17e01e80b063)

laco-dev commented on app/src/main/java/woowacourse/movie/ui/booking/presenter/BookingPresenter.kt:
```diff
@@ -76,6 +74,11 @@ class BookingPresenter(
 
     override fun completeBooking() {
-        bookingView.startBookingSeatActivity(movie.title, selectedDateTime, headcount)
+        bookingView.startBookingSeatActivity(
+            theater.movieSchedules[0].movie.title,
+            selectedDateTime,
+            headcount,
+            theater,
+        )
```
액티비티와 같이 구체적으로 View가 알아야 할 정보를 인터페이스로 남기면 안됩니다.

가령 View에서 구현을 이렇게 하더라도 전혀 문제가 없습니다.
자기가 할 일을 했기 때문이죠. 특정 구현에 의존하도록 이름을 짓는 대신 기능 관점에서 이름을 작명해주세요
```kotlin
override fun startBookingSeatActivity() {
  Toast.showToast("예약 화면 실행했음!")
}
```

laco-dev commented on app/src/main/java/woowacourse/movie/ui/complete/BookingCompleteActivity.kt:
```diff
@@ -51,17 +51,19 @@ class BookingCompleteActivity :
             movieNameTextView.text = movieName
             headcountTextView.text =
-                getString(R.string.text_headcount_with_seats).format(
+                getString(R.string.text_headcount_with_seats_and_theater).format(
```
getString()에서 어떤 파라미터를 받을 수 있는지 찾아보세요 :)

yrsel replied:
`getString()`는 `formatArgs` 도 받을 수 있어 함수에 포맷에 맞는 데이터들을 넣도록 수정하였습니다

```kotlin
binding.tvDetailInfo.text =
            getString(
                R.string.text_headcount_with_seats_and_theater,
                count, seatsNames, theaterName,
            )
```

laco-dev commented on app/src/main/java/woowacourse/movie/ui/movielist/contract/TheaterBottomSheetDialogContract.kt:
```diff
@@ -0,0 +1,16 @@
+
+interface TheaterBottomSheetDialogContract {
+    interface Presenter {
+        fun loadEntireTheaters(): Theaters
```
마찬가지로 반환하지 않도록 만들어주세요

yrsel replied:
https://github.com/woowacourse/android-movie-theater/pull/112#discussion_r2076808639

laco-dev commented on app/src/main/java/woowacourse/movie/ui/movielist/view/MovieListFragment.kt:
```diff
@@ -0,0 +1,49 @@
+    private lateinit var binding: FragmentMovieListBinding
+    private val movieListPresenter = MovieListPresenter(this)
+
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?,
+    ): View? {
+        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_movie_list, container, false)
+        movieListPresenter.loadMovieList()
+        return binding.root
+    }
```
onCreateView와 onViewCreate는 어떤 차이가 있는지 알려주세요.

yrsel replied:
`onCreateView` : 프래그먼트에 UI를 그리는 작업을 수행하는 생명주기로 Fragment와 xml을 inflate해서 화면을 연결하는 작업을 수행할 수 있습니다.
`onViewCreate` : onCreateView에서 그려진 뷰를 받아 데이터를 넣을 수 있는 생명주기로 데이터의 초기값을 넣어주는 작업 또는 리사이클러뷰 어댑터 설정 등을 수행할 수 있습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/ui/movielist/view/MovieListFragment.kt:
```diff
@@ -0,0 +1,49 @@
+        startActivity(BookingActivity.newIntent(requireActivity(), theater))
+    }
+
+    override fun setMoveListItems(items: List<MovieListItem>) {
+        val adapter =
+            MovieAdapter(
+                onClickBooking = { movie ->
+                    val theaterFragment = TheaterBottomSheetDialogFragment.newInstance(movie)
+                    theaterFragment.show(childFragmentManager, "dialog")
+                },
+            )
+
+        binding.moviesRecyclerView.adapter = adapter
+        adapter.submitList(items)
+    }
```
무한 페이지와 같이 영화 목록을 계속해서 보여주는 것을 상상해보세요.

setMovieListItems가 반복해서 호출되면 어떤 문제가 발생할까요?
```kotlin
setMovieListItems(...)
setMovieListItems(...)
setMovieListItems(...)
setMovieListItems(...)
setMovieListItems(...)
```

yrsel replied:
리사이클러뷰 어댑터도 불필요하게 매번 재생성될 것 같습니다.
리사이클러뷰 어댑터는 `onViewCreated()` 생명주기에서 한 번만 생성될 수 있도록 수정하였습니다 !

laco-dev commented on app/src/main/java/woowacourse/movie/ui/movielist/view/TheaterAdapter.kt:
```diff
@@ -0,0 +1,46 @@
+        return TheaterViewHolder(itemBinding, onClickTheater)
+    }
+
+    override fun onBindViewHolder(
+        holder: RecyclerView.ViewHolder,
+        position: Int,
+    ) {
+        val item = getItem(position)
+        (holder as TheaterViewHolder).bind(item)
+    }
```
ViewHolder 제네릭 타입에 무엇을 넣을 수 있는지 ListAdapter의 다른 사용 사례를 찾아보세요.

yrsel replied:
`RecyclerView.ViewHolder 추상 클래스`를 상속받은 `TheaterViewHolder` 타입을 제네릭 타입으로 받는다면 타입 캐스팅이 필요없어지는 이점이 있을 것 같은데 이 부분을 의도 하신 게 맞을까요 ?? 🤔

관련 커밋 fde9a5f9614fe640fdca7567789249277582a1ca

laco-dev replied:
네, 맞습니다~

laco-dev commented on app/src/main/java/woowacourse/movie/ui/movielist/view/TheaterBottomSheetDialogFragment.kt:
```diff
@@ -0,0 +1,90 @@
+    ) {
+        super.onViewCreated(view, savedInstanceState)
+
+        val movie = restoreMovie()
+        presenter.loadAvailableTheaters(movie)
```
Presenter를 lazy하게 초기화 할 수 있다면 불필요한 함수 대신 생성자로 활용할 수 있지 않을까요?

laco-dev commented on app/src/main/java/woowacourse/movie/ui/movielist/view/TheaterBottomSheetDialogFragment.kt:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    companion object {
+        @JvmStatic
```
`@JvmStatic`은 무엇인가요?

yrsel replied:
현재 상황에서 `@JvmStatic`없이 자바로 디컴파일하면 Companion 클래스에 접근해서 해당 메서드를 사용할 수 있습니다. 
반대로 `@JvmStatic`을 붙이게 된다면 자바의 static 키워드와 동일하게 사용할 수 있도록 해줍니다.

laco-dev replied:
지금 구현에서 필요한 어노테이션인가요~!?

yrsel replied:
IDE를 통해 Fragment를 생성하니 제공해준 기본 틀에 포함된 어노테이션으로 내부적으로 자바로 디컴파일할 때 필요하다고 생각했습니다.
자바, 코틀린 두 언어로 동시에 개발하고 있지 않다면 개발자 입장에서는 없어도 되는 어노테이션 인 것 같습니다 !

laco-dev commented on app/src/main/java/woowacourse/movie/ui/movielist/view/TheaterBottomSheetDialogFragment.kt:
```diff
@@ -0,0 +1,90 @@
+    ) {
+        if (theater.movieSchedules.isNotEmpty()) {
+            startActivity(BookingActivity.newIntent(view.context, theater))
+            parentFragmentManager.commit {
+                remove(this@TheaterBottomSheetDialogFragment)
+            }
```
BottomSheetDialogFragment를 종료하기 위한 인터페이스가 있습니다. 무엇일까요?

yrsel replied:
`dismiss()` 활용해서 종료하도록 수정하였습니다 ! 

관련 커밋 3f5b7361ae302a143a0798074e2a1cc7f5bf0db6

laco-dev commented on app/src/main/java/woowacourse/movie/ui/movielist/view/TheaterBottomSheetDialogFragment.kt:
```diff
@@ -0,0 +1,90 @@
+        return movie
+    }
+
+    private fun startBookingActivity(
+        view: View,
+        theater: Theater,
+    ) {
```
view가 필요한건가요?

yrsel replied:
binding 을 활용할 수 있어 제거하였습니다 ! 

관련 커밋 14341ca6799a70c3cd8039688ed0fe6f95d1d804

laco-dev commented on app/src/main/java/woowacourse/movie/ui/movielist/view/TheaterViewHolder.kt:
```diff
@@ -0,0 +1,23 @@
+    private val itemBinding: TheaterItemBinding,
+    val onClickTheater: (Theater) -> Unit,
+) : RecyclerView.ViewHolder(itemBinding.root) {
+    private var currentTheater: Theater? = null
+
+    init {
+        itemBinding.layoutTheaterItem.setOnClickListener {
+            currentTheater?.let { onClickTheater(it) }
+        }
+    }
```
bind 함수에서 onClick 리스너를 등록하면 되는데 굳이 생성자에서 불필요한 로직을 사용해가면서 만들 이유가 있나요?

yrsel replied:
`bind` 함수 내에서 리스너를 등록하게 된다면 `bind 함수` 호출 수만큼의 리스너가 등록되지만 생성자블럭에서 생성할 경우 `ViewHolder가 생성되는 시점`에만 등록되기에 효율적이라고 생각했습니다.

다만, `null` 처리 후 데이터를 가져왔는 지 확인이 필요하다보니 적은 수의 데이터인 상황 + 불필요한 로직을 줄이고자 `bind() 함수내`로 리스너 등록을 옮겼었습니다.
이후 `databinding`을 활용한다면 초기화 블럭에서 null 처리 필요없이 리스너를 등록할 수 있다는 사실을 알게되어 databinding을 활용해 리스너를 등록하는 방식으로 최종 수정하였습니다.

관련 커밋 
- TheaterAdapter (7c5b329daeea48bf2e1257f6890f2880a471b84b)
- MovieAdapter (cb11ed6f3241d963ff96a348248c3be908efe94e)

laco-dev commented on app/src/main/res/layout/fragment_movie_list.xml:
```diff
@@ -0,0 +1,34 @@
+            tools:itemCount="10"
+            tools:listitem="@layout/movie_item" />
+
+        <androidx.fragment.app.FragmentContainerView
+            android:id="@+id/fragment_theater_select_view"
+            android:layout_width="match_parent"
+            android:layout_height="250dp"
+
+            app:layout_constraintBottom_toBottomOf="parent"
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintStart_toStartOf="parent" />
```
이 View는 어떤 용도인가요?

yrsel replied:
`BottomSheetDialogFragment`를 배치할 공간을 직접 지정해줘야 하는 줄 알고 선언해놓았던 `View`였습니다. 
사용되지 않고 있어 제거하였습니다 ! 

관련 커밋 7d63e233685e9803846a29b8f8b567e8ec02d8c4

laco-dev commented on app/src/main/res/layout/fragment_booking_history.xml:
```diff
@@ -0,0 +1,13 @@
+    <TextView
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"
+        android:text="예매 내역" />
```
어디는 string resource 사용하고 어디는 하드 코딩 되어서 아쉽네요

yrsel replied:
`tools:text` 로 변경해서 개발중에만 확인할 수 있도록 수정하였습니다.

관련 커밋 6265f3604986244862cf743d93271bbc3a9f95f4

laco-dev reviewed (changes requested):
고생 많으셨습니다.
데이터바인딩 활용 방법을 잘 찾아보셨더군요. 👍🏻 

도메인 보다는 안드로이드에 친해지는 것이 중요한 단계이지만,
스스로 본문에 남기신 것과 같이 도메인 설계가 많이 어색하게 느껴집니다.

지난 미션에서 이어져온 MVP 관련 문제점들에 대해서도 남겨두었으니 피드백 받은 부분만 반영하지 말고 전반적으로 비슷한 문제점들을 같이 챙겨주세요

yrsel committed:
chore: 포스터 파일 추가

yrsel committed:
refactor: Movie id 필드 추가

yrsel committed:
chore: 리팩토링 시 제거된 코드 제거

- DayType.kt
- ScreeningDateTime.kt
- ScreeningTimeItems.kt
- ScreeningTimeItemsTest.kt

yrsel committed:
chore: Theaters 사용 안하는 코드 제거

yrsel committed:
refactor: Theater 영화따라 상영 가능 개수 반환

yrsel committed:
feat: TheaterSchedules 영화관마다 영화에 대한 스케줄 관리

yrsel committed:
refactor: MovieSchedule 영화 시작 시간과 좌석에 대한 정보

- 시간 정보만으로 동등성 비교

yrsel committed:
chore: Headcount 사용 안하는 코드 제거

yrsel committed:
refactor: 샘플 극장 데이터 구조 변경

yrsel committed:
test: 리팩토링 시 제거된 코드 테스트 코드 제거

- DayTypeTest.kt
- ScreeningDateTimeTest.kt
- TheatersTest.kt

yrsel committed:
test: 테스트 픽스처 코드 구조 변경

yrsel committed:
chore: KtlintFormat 적용

yrsel committed:
test: Fixture 패키지 분리

yrsel committed:
test: TheaterTest 테스트 코드 수정

yrsel committed:
test: MovieScheduleTest 테스트 코드 추가

yrsel committed:
test: TheaterSchedulesTest 테스트 코드 추가

yrsel committed:
refactor: Advertisement에서 광고 주기 판단

yrsel committed:
chore: 샘플 데이터 클래스 분리

yrsel committed:
refactor: TheaterSchedules 상영 가능한 영화의 개수만 출력

yrsel committed:
refactor: MovieAdapter 영화 id 값만 활용하도록 변경

yrsel committed:
refactor: MovieAdapter 영화 id 값만 활용하도록 변경

yrsel committed:
refactor: MovieListFragment MVP 구조 정리 및 반환값 제거

yrsel committed:
refactor: TheaterSchedules 직렬화 추가

yrsel committed:
refactor: TheaterAdapter 영화 id 활용하도록 변경

yrsel committed:
refactor: TheaterBottomSheetDialogFragment MVP 구조 정리 및 반환값 제거

yrsel committed:
chore: KtlintFormat 적용

yrsel committed:
feat: 영화 날짜 스피너 클래스 분리

yrsel committed:
feat: 영화 시간 스피너 클래스 분리

yrsel committed:
refactor: TheaterSchedules 날짜, 시간 데이터 정렬 후 반환하도록 수정

yrsel committed:
refactor: MovieSchedule 당일에 대한 상영 판단 로직 오류 수정

yrsel committed:
refactor: TheaterSchedules 중복기능 메서드로 추출

yrsel committed:
chore: 사용안하는 메서드 제거 및 KtlintFormat 적용

yrsel committed:
refactor: BookingActivity MVP 구조 정리 및 반환값 제거

yrsel committed:
refactor: BookingSeatActivity data binding 적용

yrsel committed:
refactor: BookingSeatActivity MVP 구조 정리 및 반환값 제거

yrsel committed:
refactor: Seats 기존 좌석 고려할 수 있도록 구성

yrsel committed:
refactor: BookedTicket 내부 구성요소 변경

yrsel committed:
refactor: BookingCompleteActivity MVP 구조 정리 및 반환값 제거

yrsel committed:
test: SeatsTest 테스트 코드 통과하도록 수정

yrsel committed:
test: TheaterTest 테스트 코드 통과하도록 수정

yrsel committed:
test: BookingActivityTest 테스트 코드 통과하도록 수정

yrsel committed:
test: BookingCompleteActivityTest 테스트 코드 통과하도록 수정

yrsel committed:
test: BookingSeatActivityTest 테스트 코드 통과하도록 수정

- BookingSeatPresenter 누락된 버튼 클릭 가능 여부 로직 추가

yrsel committed:
test: MovieListPresenterTest 테스트 코드 통과하도록 수정

yrsel committed:
test: TheaterBottomSheetDialogPresenterTest 테스트 코드 통과하도록 수정

yrsel committed:
test: BookingPresenterTest 테스트 코드 통과하도록 수정

yrsel committed:
test: BookingSeatPresenterTest 테스트 코드 통과하도록 수정

yrsel committed:
test: BookingCompletePresenterTest 테스트 코드 통과하도록 수정

yrsel committed:
chore: KtlintFormat 적용

yrsel committed:
refactor: MovieBookingActivity 바텀 네비 홈 고정 문제 수정

yrsel committed:
refactor: MovieBookingActivity 프래그먼트들 인스턴스화 수정

yrsel committed:
refactor: TheaterBottomSheetDialogFragment 종료 메서드 호출하도록 수정

yrsel committed:
chore: 개발 확인용 텍스트 tools name space 로 변경

yrsel committed:
chore: 주석 제거

yrsel committed:
chore: MovieListFragment xml에 사용되지 않는 View 제거

yrsel committed:
chore: 사용되지 않는 클래스 제거 및 KtlintFormat 적용

- 제거 : AdapterItemSelectedListener.kt

yrsel committed:
refactor: TheaterAdapter data binding으로 리스너 설정

yrsel committed:
refactor: MovieAdapter data binding으로 리스너 설정

yrsel committed:
refactor: TheaterAdapter ViewHolder 타입 변경

yrsel commented:
@coderabbitai pause

laco-dev reviewed (approved):
피드백 굉장히 꼼꼼하게 남겨주셨네요 고맙습니다. 👍🏻 
충분히 고민하고 연습해보셨으니 남은 시간은 3, 4단계에 집중해보면 좋을 것 같아요.

마지막까지 화이팅입니다 :)
